### ISSUE TITLE
Support Component property annotations as defined by DS 1.4

### ISSUE THREAD

**Issue by timothyjward** [bndtools/bnd#2163]
These annotations are used to provide type-safe property values for DS components, and are meta annotated with @ComponentProperty

**Comment by kwin** [comment]
@timothyjward @bjhargrave AFAIK the component property annotations are a compile-time only specification. At run-time everything looks as before. So what is the reason to always emit DS 1.4 component descriptors when annotations have been used (https://github.com/timothyjward/bnd/blob/e8eadbde655f6e4c100aa7b01ac1621f93045402/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L515)

**Comment by bjhargrave** [comment]
It does not emit 1.4 currently:

https://github.com/bndtools/bnd/blob/f36876406eb8fd4adae71c414537448a67b5ec80/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L558-L560

**Comment by timothyjward** [comment]
Also - there is a difference between using a non-annotated Component Property Type (DS 1.3) and the `@ComponentPropertyType` meta annotation. The former does not imply DS 1.4, but the latter does.

**Comment by kwin** [comment]
The workaround is not viable. Basically the use case is to provide Custom annotations for Apache Sling. Those must work with DS 1.3. What is the reason why the capability being generated from bnd is osgi.component 1.4.0 or newer in this case?

**Comment by timothyjward** [comment]
As I mentioned earlier up the chain, the 1.4 version of `@Component` Is annotated with an `@Requirement` for the DS 1.4 extender. If you have the DS 1.4 annotations ahead of the DS 1.3 annotations on your build class path then it will be the one bnd finds and you will get that requirement added. 

If you make sure that the DS 1.3 annotations are ahead of the 1.4 annotations on your build path then that should prevent the 1.4 extender requirement being added, and put a 1.3 extender requirement instead.

**Comment by kwin** [comment]
> According to @rotty3000 it’s possible to override generated capabilities in the bnd file. I think you can repeat the extender requirement, but make it optional by adding resolution:=optional which will work around your problem.

I tried that by adding `Require-Capability: osgi.extender;filter:="(&(osgi.extender=osgi.component)(&(version>=1.4.0)(!(version>=2.0.0))))";resolution:=optional` to my bnd file. Now the generated header looks like this

```Require-Capability: osgi.extender; filter:="(&(osgi.extender=osgi.component)(&(version>=1.4.0)(!(version>=2.0.0))))"; resolution:=optional, osgi.extender; filter:="(&(osgi.extender=osgi.component)(version>=1.3.0)(!(version>=2.0.0)))", osgi.extender; filter:="(&(osgi.extender=osgi.component)(&(version>=1.4.0)(!(version>=2.0.0))))", osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=1.8))"```
This will obviously still not resolve yet!

Putting the DS 1.3 annotations in the classpath as well is not that easy with Maven.

**Comment by cvgaviao** [comment]
I'm facing a similar problem while trying bndlib 4.0 with maven-bundle-plugin. But in my case there a duplication of requirements:
` Require-Capability: osgi.extender;filter:="(&(osgi.extender=osgi.component)(version>=1.3.0)(!(version>=2.0.0)))",osgi.service;filter:="(objectClass=org.osgi.service.component.ComponentFactory)";effective:=active,osgi.extender;filte
 r:="(&(osgi.extender=osgi.component)(&(version>=1.4.0)(!(version>=2.0
 .0))))",osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
`

**Comment by pkriens** [comment]
I do not see a duplication?

    Require-Capability: 
        osgi.extender;filter:="(&
              (osgi.extender=osgi.component)
              (version>=1.3.0)(!(version>=2.0.0)))",
       osgi.service;filter:=" 
           (objectClass=org.osgi.service.component.ComponentFactory)";
           effective:=active,
      osgi.extender;filter:="(&
             (osgi.extender=osgi.component)
             (&(version>=1.4.0)(!(version>=2.0.0))))",
      osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"

Those are 4 distinct requirements?

**Comment by cvgaviao** [comment]
@pkriens, sorry I didn't see your comment in my mailbox. 
There are two requirements for osgi.extender=osgi.component, but one is for version 1.3 and the other for 1.4. Both are mandatory.

**Comment by pkriens** [comment]
That is perfectly ok because use a range up to 2.0.0. This is perfectly normal.

**Comment by cvgaviao** [comment]
Well, but even not using directly an annotation from DS 1.4. 
In may case, I changed the osgi.core, osgi.cmpn and osgi.annotations bundles to version 7.0 in order to access other things from R7 other than DS 1.4. But doing so, is prevent us to run our bundles with Equinox, since it provides DS 1.3.

**Comment by cvgaviao** [comment]
@pkriens, perhaps would be possible to introduce a property where we indicate  which version of DS will be written in the manifest ?

**Comment by pkriens** [comment]
When DS generates the XML you can set the version:

      -dsannotations-options                :         version;minimum=1.2.0

You can control the maximum version by (not) using the features of that version.

**Comment by cvgaviao** [comment]
@pkriens, correct if I'm wrong...  `-dsannotations-options ` will interfere only in the generated XML but not in the generated Require-Capability, right ?

As it was already said above just the fact to have the osgi.core 7.0.0 on the classpath will drive bndlib to use 1.4.0  
But that is not good, even if we are using only @Component in our components, those generated requirements will prevent us to use other good things from R7 spec: aka Converter, Configurator, etc...

**Comment by bjhargrave** [comment]
> osgi.core 7.0.0

I assume you meant osgi.cmpn.

OSGi also published individual jars for specs if you want to use some new R7 specs and osgi.cmpn-6.0.0.

**Comment by cvgaviao** [comment]
@bjhargrave, I will see tomorrow how that individual jars mixed with cmpn-6.0.0 would behave in a maven based project.

**Comment by cvgaviao** [comment]
Hi @bjhargrave, I have done some experiments with the individual jars that you have cited.

I've removed the osgi.cmpn dependency and then added all the individual spec jars from R7 that I'm using. For DS, I have added `org.osgi.service.component:1.3.0` and `org.osgi.service.component.annotations:1.3.0`.  

This was enough for bndlib to generate the manifest properly, without the requirement on DS-1.4.0 and I was able to run my bundles properly inside both equinox and felix. 

Also, one good thing is that I was able to at least use the annotations `@Capability` and `@Requirement`, and now I do not need to write capabilities and requirements by hand in my pom.xml.

Thanks all
